
%% bare_jrnl_transmag.tex
%% V1.4a
%% 2014/09/17
%% by Michael Shell
%% see http://www.michaelshell.org/
%% for current contact information.
%%
%% This is a skeleton file demonstrating the use of IEEEtran.cls
%% (requires IEEEtran.cls version 1.8a or later) with an IEEE 
%% Transactions on Magnetics journal paper.
%%
%% Support sites:
%% http://www.michaelshell.org/tex/ieeetran/
%% http://www.ctan.org/tex-archive/macros/latex/contrib/IEEEtran/
%% and
%% http://www.ieee.org/

%%*************************************************************************
%% Legal Notice:
%% This code is offered as-is without any warranty either expressed or
%% implied; without even the implied warranty of MERCHANTABILITY or
%% FITNESS FOR A PARTICULAR PURPOSE! 
%% User assumes all risk.
%% In no event shall IEEE or any contributor to this code be liable for
%% any damages or losses, including, but not limited to, incidental,
%% consequential, or any other damages, resulting from the use or misuse
%% of any information contained here.
%%
%% All comments are the opinions of their respective authors and are not
%% necessarily endorsed by the IEEE.
%%
%% This work is distributed under the LaTeX Project Public License (LPPL)
%% ( http://www.latex-project.org/ ) version 1.3, and may be freely used,
%% distributed and modified. A copy of the LPPL, version 1.3, is included
%% in the base LaTeX documentation of all distributions of LaTeX released
%% 2003/12/01 or later.
%% Retain all contribution notices and credits.
%% ** Modified files should be clearly indicated as such, including  **
%% ** renaming them and changing author support contact information. **
%%
%% File list of work: IEEEtran.cls, IEEEtran_HOWTO.pdf, bare_adv.tex,
%%                    bare_conf.tex, bare_jrnl.tex, bare_conf_compsoc.tex,
%%                    bare_jrnl_compsoc.tex, bare_jrnl_transmag.tex
%%*************************************************************************


% *** Authors should verify (and, if needed, correct) their LaTeX system  ***
% *** with the testflow diagnostic prior to trusting their LaTeX platform ***
% *** with production work. IEEE's font choices and paper sizes can       ***
% *** trigger bugs that do not appear when using other class files.       ***                          ***
% The testflow support page is at:
% http://www.michaelshell.org/tex/testflow/



\documentclass[journal,transmag,twoside]{IEEEtran}
%
% If IEEEtran.cls has not been installed into the LaTeX system files,
% manually specify the path to it like:
% \documentclass[journal]{../sty/IEEEtran}





% Some very useful LaTeX packages include:
% (uncomment the ones you want to load)


% *** MISC UTILITY PACKAGES ***
%
%\usepackage{ifpdf}
% Heiko Oberdiek's ifpdf.sty is very useful if you need conditional
% compilation based on whether the output is pdf or dvi.
% usage:
% \ifpdf
%   % pdf code
% \else
%   % dvi code
% \fi
% The latest version of ifpdf.sty can be obtained from:
% http://www.ctan.org/tex-archive/macros/latex/contrib/oberdiek/
% Also, note that IEEEtran.cls V1.7 and later provides a builtin
% \ifCLASSINFOpdf conditional that works the same way.
% When switching from latex to pdflatex and vice-versa, the compiler may
% have to be run twice to clear warning/error messages.






% *** CITATION PACKAGES ***
%
%\usepackage{cite}
% cite.sty was written by Donald Arseneau
% V1.6 and later of IEEEtran pre-defines the format of the cite.sty package
% \cite{} output to follow that of IEEE. Loading the cite package will
% result in citation numbers being automatically sorted and properly
% "compressed/ranged". e.g., [1], [9], [2], [7], [5], [6] without using
% cite.sty will become [1], [2], [5]--[7], [9] using cite.sty. cite.sty's
% \cite will automatically add leading space, if needed. Use cite.sty's
% noadjust option (cite.sty V3.8 and later) if you want to turn this off
% such as if a citation ever needs to be enclosed in parenthesis.
% cite.sty is already installed on most LaTeX systems. Be sure and use
% version 5.0 (2009-03-20) and later if using hyperref.sty.
% The latest version can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/cite/
% The documentation is contained in the cite.sty file itself.






% *** GRAPHICS RELATED PACKAGES ***
%
\ifCLASSINFOpdf
  \usepackage[pdftex]{graphicx}
  % declare the path(s) where your graphic files are
  % \graphicspath{{../pdf/}{../jpeg/}}
  % and their extensions so you won't have to specify these with
  % every instance of \includegraphics
  % \DeclareGraphicsExtensions{.pdf,.jpeg,.png}
\else
  % or other class option (dvipsone, dvipdf, if not using dvips). graphicx
  % will default to the driver specified in the system graphics.cfg if no
  % driver is specified.
  % \usepackage[dvips]{graphicx}
  % declare the path(s) where your graphic files are
  % \graphicspath{{../eps/}}
  % and their extensions so you won't have to specify these with
  % every instance of \includegraphics
  % \DeclareGraphicsExtensions{.eps}
\fi
% graphicx was written by David Carlisle and Sebastian Rahtz. It is
% required if you want graphics, photos, etc. graphicx.sty is already
% installed on most LaTeX systems. The latest version and documentation
% can be obtained at: 
% http://www.ctan.org/tex-archive/macros/latex/required/graphics/
% Another good source of documentation is "Using Imported Graphics in
% LaTeX2e" by Keith Reckdahl which can be found at:
% http://www.ctan.org/tex-archive/info/epslatex/
%
% latex, and pdflatex in dvi mode, support graphics in encapsulated
% postscript (.eps) format. pdflatex in pdf mode supports graphics
% in .pdf, .jpeg, .png and .mps (metapost) formats. Users should ensure
% that all non-photo figures use a vector format (.eps, .pdf, .mps) and
% not a bitmapped formats (.jpeg, .png). IEEE frowns on bitmapped formats
% which can result in "jaggedy"/blurry rendering of lines and letters as
% well as large increases in file sizes.
%
% You can find documentation about the pdfTeX application at:
% http://www.tug.org/applications/pdftex




% *** MATH PACKAGES ***
%
%\usepackage[cmex10]{amsmath}
% A popular package from the American Mathematical Society that provides
% many useful and powerful commands for dealing with mathematics. If using
% it, be sure to load this package with the cmex10 option to ensure that
% only type 1 fonts will utilized at all point sizes. Without this option,
% it is possible that some math symbols, particularly those within
% footnotes, will be rendered in bitmap form which will result in a
% document that can not be IEEE Xplore compliant!
%
% Also, note that the amsmath package sets \interdisplaylinepenalty to 10000
% thus preventing page breaks from occurring within multiline equations. Use:
%\interdisplaylinepenalty=2500
% after loading amsmath to restore such page breaks as IEEEtran.cls normally
% does. amsmath.sty is already installed on most LaTeX systems. The latest
% version and documentation can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/required/amslatex/math/





% *** SPECIALIZED LIST PACKAGES ***
%
%\usepackage{algorithmic}
% algorithmic.sty was written by Peter Williams and Rogerio Brito.
% This package provides an algorithmic environment fo describing algorithms.
% You can use the algorithmic environment in-text or within a figure
% environment to provide for a floating algorithm. Do NOT use the algorithm
% floating environment provided by algorithm.sty (by the same authors) or
% algorithm2e.sty (by Christophe Fiorio) as IEEE does not use dedicated
% algorithm float types and packages that provide these will not provide
% correct IEEE style captions. The latest version and documentation of
% algorithmic.sty can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/algorithms/
% There is also a support site at:
% http://algorithms.berlios.de/index.html
% Also of interest may be the (relatively newer and more customizable)
% algorithmicx.sty package by Szasz Janos:
% http://www.ctan.org/tex-archive/macros/latex/contrib/algorithmicx/




% *** ALIGNMENT PACKAGES ***
%
%\usepackage{array}
% Frank Mittelbach's and David Carlisle's array.sty patches and improves
% the standard LaTeX2e array and tabular environments to provide better
% appearance and additional user controls. As the default LaTeX2e table
% generation code is lacking to the point of almost being broken with
% respect to the quality of the end results, all users are strongly
% advised to use an enhanced (at the very least that provided by array.sty)
% set of table tools. array.sty is already installed on most systems. The
% latest version and documentation can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/required/tools/


% IEEEtran contains the IEEEeqnarray family of commands that can be used to
% generate multiline equations as well as matrices, tables, etc., of high
% quality.




% *** SUBFIGURE PACKAGES ***
%\ifCLASSOPTIONcompsoc
%  \usepackage[caption=false,font=normalsize,labelfont=sf,textfont=sf]{subfig}
%\else
%  \usepackage[caption=false,font=footnotesize]{subfig}
%\fi
% subfig.sty, written by Steven Douglas Cochran, is the modern replacement
% for subfigure.sty, the latter of which is no longer maintained and is
% incompatible with some LaTeX packages including fixltx2e. However,
% subfig.sty requires and automatically loads Axel Sommerfeldt's caption.sty
% which will override IEEEtran.cls' handling of captions and this will result
% in non-IEEE style figure/table captions. To prevent this problem, be sure
% and invoke subfig.sty's "caption=false" package option (available since
% subfig.sty version 1.3, 2005/06/28) as this is will preserve IEEEtran.cls
% handling of captions.
% Note that the Computer Society format requires a larger sans serif font
% than the serif footnote size font used in traditional IEEE formatting
% and thus the need to invoke different subfig.sty package options depending
% on whether compsoc mode has been enabled.
%
% The latest version and documentation of subfig.sty can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/subfig/



% *** FLOAT PACKAGES ***
%
%\usepackage{fixltx2e}
% fixltx2e, the successor to the earlier fix2col.sty, was written by
% Frank Mittelbach and David Carlisle. This package corrects a few problems
% in the LaTeX2e kernel, the most notable of which is that in current
% LaTeX2e releases, the ordering of single and double column floats is not
% guaranteed to be preserved. Thus, an unpatched LaTeX2e can allow a
% single column figure to be placed prior to an earlier double column
% figure. The latest version and documentation can be found at:
% http://www.ctan.org/tex-archive/macros/latex/base/


%\usepackage{stfloats}
% stfloats.sty was written by Sigitas Tolusis. This package gives LaTeX2e
% the ability to do double column floats at the bottom of the page as well
% as the top. (e.g., "\begin{figure*}[!b]" is not normally possible in
% LaTeX2e). It also provides a command:
%\fnbelowfloat
% to enable the placement of footnotes below bottom floats (the standard
% LaTeX2e kernel puts them above bottom floats). This is an invasive package
% which rewrites many portions of the LaTeX2e float routines. It may not work
% with other packages that modify the LaTeX2e float routines. The latest
% version and documentation can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/sttools/
% Do not use the stfloats baselinefloat ability as IEEE does not allow
% \baselineskip to stretch. Authors submitting work to the IEEE should note
% that IEEE rarely uses double column equations and that authors should try
% to avoid such use. Do not be tempted to use the cuted.sty or midfloat.sty
% packages (also by Sigitas Tolusis) as IEEE does not format its papers in
% such ways.
% Do not attempt to use stfloats with fixltx2e as they are incompatible.
% Instead, use Morten Hogholm'a dblfloatfix which combines the features
% of both fixltx2e and stfloats:
%
% \usepackage{dblfloatfix}
% The latest version can be found at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/dblfloatfix/




%\ifCLASSOPTIONcaptionsoff
%  \usepackage[nomarkers]{endfloat}
% \let\MYoriglatexcaption\caption
% \renewcommand{\caption}[2][\relax]{\MYoriglatexcaption[#2]{#2}}
%\fi
% endfloat.sty was written by James Darrell McCauley, Jeff Goldberg and 
% Axel Sommerfeldt. This package may be useful when used in conjunction with 
% IEEEtran.cls'  captionsoff option. Some IEEE journals/societies require that
% submissions have lists of figures/tables at the end of the paper and that
% figures/tables without any captions are placed on a page by themselves at
% the end of the document. If needed, the draftcls IEEEtran class option or
% \CLASSINPUTbaselinestretch interface can be used to increase the line
% spacing as well. Be sure and use the nomarkers option of endfloat to
% prevent endfloat from "marking" where the figures would have been placed
% in the text. The two hack lines of code above are a slight modification of
% that suggested by in the endfloat docs (section 8.4.1) to ensure that
% the full captions always appear in the list of figures/tables - even if
% the user used the short optional argument of \caption[]{}.
% IEEE papers do not typically make use of \caption[]'s optional argument,
% so this should not be an issue. A similar trick can be used to disable
% captions of packages such as subfig.sty that lack options to turn off
% the subcaptions:
% For subfig.sty:
% \let\MYorigsubfloat\subfloat
% \renewcommand{\subfloat}[2][\relax]{\MYorigsubfloat[]{#2}}
% However, the above trick will not work if both optional arguments of
% the \subfloat command are used. Furthermore, there needs to be a
% description of each subfigure *somewhere* and endfloat does not add
% subfigure captions to its list of figures. Thus, the best approach is to
% avoid the use of subfigure captions (many IEEE journals avoid them anyway)
% and instead reference/explain all the subfigures within the main caption.
% The latest version of endfloat.sty and its documentation can obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/endfloat/
%
% The IEEEtran \ifCLASSOPTIONcaptionsoff conditional can also be used
% later in the document, say, to conditionally put the References on a 
% page by themselves.




% *** PDF, URL AND HYPERLINK PACKAGES ***
%
\usepackage{url}
% url.sty was written by Donald Arseneau. It provides better support for
% handling and breaking URLs. url.sty is already installed on most LaTeX
% systems. The latest version and documentation can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/url/
% Basically, \url{my_url_here}.

\usepackage{hyperref}
\newcommand{\email}[1]{\href{mailto:#1}{#1}}

\usepackage{pdfcomment}
\newcommand{\karrcomment}[2]{\pdfmarkupcomment[markup=Highlight,color=yellow,author={Jonathan Karr}]{#1}{#2}}
\usepackage[pdftex,rgb,dvipsnames,svgnames,hyperref,table]{xcolor}
\newcommand{\change}[1]{\textcolor{DarkBlue}{#1}}
\usepackage{pgfplots}
\usepackage{caption}
\usepackage{subcaption}
\usepackage{fancyvrb}

% *** Do not adjust lengths that control margins, column widths, etc. ***
% *** Do not use packages that alter fonts (such as pslatex).         ***
% There should be no need to do such things with IEEEtran.cls V1.6 and later.
% (Unless specifically asked to do so by the journal or conference you plan
% to submit to, of course. )

\usepackage{color,soul} % for highlighting only


%http://tex.stackexchange.com/questions/61265/why-doesnt-the-ieeetran-document-class-recognize-definition-environments
\newtheorem{definition}{Definition}

% correct bad hyphenation here
\hyphenation{net-works}

% Nature News on irreproducibility:
% http://www.nature.com/news/reproducibility-1.17552

\begin{document}

\newcommand{\thetitle}{Guidelines for reproducibly building and simulating systems biology models}
\title{\change{\thetitle}}

\author{
    J. Kyle Medley$^*$ and
	Jonathan R. Karr
    
    \thanks{
        Manuscript received November 4, 2015; revised XXX XX, 2016; accepted XXX XX, 2016. Date of publication XXX XX, 2016; date of current version XXX XX, 2016.
        J. K. Medley was supported by National Institutes of Health grant R01 GM081070. J. R. Karr was supported by National Science Foundation grant 1548123 and a James S. McDonnell Foundation Postdoctoral Fellowship Award in Studying Complex Systems. The content is solely the responsibility of the authors and does not necessarily represent the views of the National Institutes of Health, the National Science Foundation, or the James S. McDonnell Foundation.
        \textit{Asterisk indicates corresponding author.}
    }
    \thanks{$^*$J. K. Medley is with the Department of Bioengineering, University of Washington, Seattle, WA 98195, USA (e-mail: \email{medleyj@uw.edu}).}
    \thanks{J. R. Karr is with the Department of Genetics \& Genomic Sciences, Icahn School of Medicine at Mount Sinai, New York, NY 10029, USA (e-mail: \email{karr@mssm.edu}).}
    \thanks{Digital Object Identifier 10.1109/TBME.XXXX.XXXXXXX}
}

% The paper headers
\markboth{IEEE Transactions on Biomedical Engineering,~Vol.~XX, No.~XX, XXXX~2016}%
{Medley and Karr: \thetitle}
% The only time the second header will appear is for the odd numbered pages
% after the title page when using the twoside option.
% 
% *** Note that you probably will NOT want to include the author's ***
% *** name in the headers of peer review papers.                   ***
% You can use \ifCLASSOPTIONpeerreview for conditional compilation here if
% you desire.




% If you want to put a publisher's ID mark on the page you can do it like
% this:
%\IEEEpubid{0000--0000/00\$00.00~\copyright~2014 IEEE}
% Remember, if you use this you must call \IEEEpubidadjcol in the second
% column for its text to clear the IEEEpubid mark.



% use for special paper notices
%\IEEEspecialpapernotice{(Invited Paper)}

\maketitle

\begin{abstract}\change{
\textit{Objective:}
Reproducibility is the cornerstone of the scientific method. 
However, currently many systems biology models cannot easily be reproduced.
\textit{Methods:} 
We analyzed the recent \textit{Mycoplasma genitalium} whole-cell model to determine the requirements for reproducible systems biology modeling.
\textit{Results:} 
We determined that reproducible modeling requires both reproducible model building and reproducible numerical simulation.
\textit{Conclusion:} 
We concluded that new standards and simulation software tools are needed to enhance the reproducibility of systems biology modeling. 
New standards are needed to explicitly document all model data source and assumptions, and new parallel simulation tools are needed to deterministically simulate complex models.
\textit{Significance:}
We anticipate that these new standards and software will enable researchers to reproducibly build and simulate more complex models.
}
\end{abstract}

\begin{IEEEkeywords}
Systems biology, whole-cell, computational modeling, reproducibility, standards
\end{IEEEkeywords}

% To allow for easy dual compilation without having to reenter the
% abstract/keywords data, the \IEEEtitleabstractindextext text will
% not be used in maketitle, but will appear (i.e., to be "transported")
% here as \IEEEdisplaynontitleabstractindextext when the compsoc 
% or transmag modes are not selected <OR> if conference mode is selected 
% - because all conference papers position the abstract like regular
% papers do.
\IEEEdisplaynontitleabstractindextext
% \IEEEdisplaynontitleabstractindextext has no effect when using
% compsoc or transmag under a non-conference mode.







% For peer review papers, you can put extra information on the cover
% page as needed:
% \ifCLASSOPTIONpeerreview
% \begin{center} \bfseries EDICS Category: 3-BBND \end{center}
% \fi
%
% For peerreview papers, this IEEEtran command inserts a page break and
% creates the second title. It will be ignored for other modes.
\IEEEpeerreviewmaketitle



\section{Introduction}
% The very first letter is a 2 line initial drop letter followed
% by the rest of the first word in caps.
% 
% form to use if the first word consists of a single letter:
% \IEEEPARstart{A}{demo} file is ....
% 
% form to use if you need the single drop letter followed by
% normal text (unknown if ever used by IEEE):
% \IEEEPARstart{A}{}demo file is ....
% 
% Some journals put the first two words in caps:
% \IEEEPARstart{T}{his demo} file is ....
% 
% Here we have the typical use of a "T" for an initial drop letter
% and "HIS" in caps to complete the first word.
\IEEEPARstart{R}{eproducibility} is one of the central tenets of the scientific method.
\textit{Reproducibility} is the ability to confirm a result via a completely independent test, including by different investigators, experimental methods, and experimental machinery. 
This rigorous standard eliminates conclusions that were based on incorrect methods, machinery, or experiments, and ensures that scientific results are only accepted as facts once multiple scientists have thoroughly dismissed other potential explanations. 

\textit{Repeatability} is the ability to reproduce a result given the same experimental machinery and conditions.
Repeatability is a more lenient standard than reproducibility. 
Consequently, repeatability only eliminates conclusions that were based on erroneous experiments, and cannot eliminate conclusions that were based on faulty methods or machinery.

\change{The following thought experiment illustrates the difference between reproducibility and repeatability. Consider a modeler Alice who wants to investigate a discrepant simulation result published by Bob about a pathway which has been implicated in cancer. Bob's model predicts that knocking out regulator Y causes cancer, whereas Alice's model indicates that cancer requires at least two knockouts. Fortunately, Alice can investigate Bob's results because Bob published his model in the Systems Biology Markup Language (SBML) \cite{hucka2003} format and his simulation experiments in the Simulation Experiment Description Markup Language (SED-ML) \cite{sedml2011} format. Alice uses Bob's SBML and SED-ML files to \textit{replicate} Bob's simulation results, compares the two models, and finds that Bob's model generates different predictions because it uses a different rate law to describe the effect of regulatory Y. However, because Bob's SBML file does not describe all of the experimental data and assumptions underlying his model, Alice cannot \textit{reproduce} Bob's rate laws and thus cannot fully resolve why their models generate different predictions.}

The systems biology community has developed several standard formats to exchange systems biology models and repeat systems biology simulations. These include CellML \cite{cuellar2003overview}, the Computational Modeling in Biology Network (COMBINE) archive \cite{COMBINE2012}, SBML \cite{hucka2003}, SED-ML \cite{sedml2011}, and the Systems Biology Graphical Notation (SBGN) \cite{LeNovereHMMSS09}. These standards support several types of systems biology models including ordinary differential equation (ODE), flux balance analysis \cite{orth2010flux}, and logical models and numerous modeling software programs support these standards \cite{hucka2011profile}. Several model repositories have also been developed to facilitate model sharing, including BioModels \cite{chelliah2015biomodels} and the CellML Model Repository \cite{lloyd2008cellml}. 

These standards and modeling software tools help researchers repeat most \textit{in silico} systems biology experiments and reuse, modify, expand, and combine most systems biology models. However, these standards do not completely help researchers reproduce models because they do not record the experimental data and assumptions underlying models.

Furthermore, researchers have begun to develop more complex models which cannot be represented by the existing systems biology standards and simulated by the existing standards-compliant software. For example, several aspects of the recent whole-cell (WC) model of the gram-positive bacterium \textit{Mycoplasma genitalium} \cite{Karr2012} cannot currently be represented in SBML \cite{Waltemath2016}. In particular, SBML cannot represent multi-algorithmic models, such as the \textit{M. genitalium} WC model, which are composed of multiple mathematically diverse submodels of individual cellular pathways.

\change{Here, we must present a comprehensive vision for reproducible systems biology modeling, including complex multi-algorithmic WC models. First, we describe the requirements for reproducible modeling. Second, we outline the gaps in the existing standards and software for reproducible modeling. Third, we describe several new standards and software tools that are needed for reproducible modeling, including new model design tools which automatically record every model source and assumption so that models can be reproduced from their underlying data and assumptions, new standards for multi-algorithmic and rule- and sequence-based modeling, and new simulation software programs which can quickly and deterministically simulate models using large computing clusters with variable available resources. Deterministic simulation is not strictly required for reproducibility. However, we believe that determinism is invaluable for debugging complex models.}

%CellML and SBML must be expanded to describe WC models. A new model design tool which automatically records every data source and assumption must also be developed so that WC models can be reproduced. In addition, a new repeatable multi-algorithm simulator must be developed. This simulator should use parallel computing to achieve adequate performance and use deterministic frameworks to generate reproducible results from variable computing resources.

%What role does \textit{repeatability} play in the process of performing reproducible research?

%Although the topic of the workshop was reproducibility, we have also chosen to include \textit{repeatability} in our discussion. We believe that, particularly for re-encoding efforts, the determinism of the original simulation is a significant factor. Consider again the above example and suppose Bob's simulation was based on a stochastic method instead of a deterministic one. Alice can still compare the propensities at the initial time point, but if the initial propensities are identical then the timecourse behavior of the model will not help her to isolate the discrepancy. Alice can employ other methods to continue trying to reproduce Bob's results but by throwing away the ability to produce predictable, deterministic simulations Bob has effectively made it difficult to reproduce his work.

%There are certainly times when the requirement of replicating simulation results exactly does not apply, but since our discussion is motivated by the drive to create a better platform for modeling complex systems in a robust way, we believe it is useful to consider whether exact repeatability aids in the development, encoding, and distribution of complex hybrid models.

% Due to the complexity of modern scientific experiments as well as financial, time, and publishing pressures, investigators often do not have the time or resources to design orthogonal methods to test results independently. Instead, investigators frequently replicate reported results using the published method. The underemphasis on independent testing has the potential to allow us to accept false results based on faulty methods.

\begin{figure*}[!t]
\centering
\includegraphics[width=\textwidth]{figure1/figure1}
% where an .eps filename suffix will be assumed under latex,
% and a .pdf suffix will be assumed for pdflatex; or what has been declared
% via \DeclareGraphicsExtensions.
\caption{Requirements for reproducing systems biology modeling. (1) Every data source and assumption underlying the model must be recorded so that the model species, reactions, and rate laws can be regenerated. (2) Every simulation parameter value must be recorded so that the numerical simulation results can be regenerated. (3) Multiple simulation software tools must generate consistent numerical results. Reproducibility is a stricter standard than repeatability. Models are considered repeatable, but not reproducible, if they satisfy the second and third criteria.}
\label{fig_repro_diagram}
\end{figure*}

\change{\section{The Requirements for Reproducible Systems Biology Modeling}}

\change{There are three requirements for reproducible systems biology modeling (Fig.~\ref{fig_repro_diagram}). First, researchers must record every data and assumption used to construct models so that other researchers can generate the same modeled species, reactions, and rate laws. Second, researchers must record every algorithm and parameter value used to simulate models so that other researchers can generate the same numerical results. Third, researchers must ensure that every simulation software tool produces statistically indistinguishable numerical results so that researchers who use different software programs can reach the same conclusions.}

\change{\section{Progress Toward Reproducible Systems Biology Modeling}}

This requires new machine-readable experimental databases which contain the data needed to build models, new model annotation standards for describing how these databases are used to build models, and new model annotation standards for describing the assumptions that models are based on \cite{boulton2012open}.

The result of the model construction steps described above will usually be a set of differential equations, reaction propensities, or rules which govern the timecourse behavior of the model. While in theory this formalism should be sufficient to recapitulate the model's behavior, we have found that reproducibility is further complicated by the details of the software implementation. In MATLAB, for example, there are a number of differential equation solvers for both stiff and non-stiff systems. These solvers in turn have many configuration options relating to error tolerances and variable stepping. Therefore, we argue that software plays a critical role in reproducibility. Making the source code available partly addresses this problem \cite{easterbrook2014open}, but we will show that more can be done to improve reproducibility in this area.

Stochastic models are problematic for reproducibility studies because they inherently give different output each time they are run. However, real implementations of stochastic processes often use pseudo-random number generators (PRNG) to achieve the illusion of stochasticity. The current standard for reproducing a stochastic model is based on comparing at least the first and second moments of the model's predictions. However, we argue that there are two advantages to reproducing the model's exact trajectory.

\karrcomment{First, biology is replete with examples of important rare events: e.g. cancer, speciation events, and mutations conferring antibiotic resistance. We believe that capturing these events is instrumental to the usefulness of WC models as engines for discovery.}{Interesting, but how is this related to repeatability?}

Second, reproducing predicted timecourses is critical to debugging complex models. The ability to reproduce exact model traces is highly useful for debugging models because it helps modelers identify the exact point where models fail.

\karrcomment{The idea of capturing the exact sequence of events leading to an error is not new.}{This is a different topic. This should be its own separate section.} In the field of computer science, the SPIN model checker \cite{holzmann1997model, holzmann2004spin} has been used extensively to validate the operational logic of non-deterministic systems. SPIN detects invalid models by exhibiting a counterexample, including the exact events leading to failure. We envisage a similar approach for biophysical modeling. Specifically, we envisage a simulator where not only can an exact simulation be replicated by supplying the same random seed, but also where a simulation may be restarted from a checkpoint containing a state snapshot at any point in the timecourse. Since serializing such checkpoints would be an expensive operation, the checkpoints would be user-specified. By combining these two features, we would be able to (1) replicate exactly a trace from a specific prior simulation, and (2) insert as many such checkpoints into the run as desired without affecting the outcome. We would then be able to capture rare events and apply permutations to test various hypotheses relating to the mechanism of the events, e.g. to test their robustness to knockouts and systematically screen for targets for intervention.

\subsection{Validation Tests as a Basis for Replicability}

Unit testing is a powerful strategy for ensuring the replicability of models. Bergmann et al. and Evans et al. have used unit testing to debug multiple SBML-compatable simulators until they produced the same ODE model predictions \cite{bergmann2008comparing, evans2008sbml}. The replicability of stochastic models, such as WC models, can be statistically assessed by comparing their predicted mean and standard deviations across multiple simulators.

However, in practice, we believe that statistical tests are insufficient to evaluate the replicability of stochastic models. \karrcomment{It has been shown that noise can have non-linearly effects on network behaviors}{I don't find this very convincing argument. Statistical tests for variance should pickup bimodality. I think the main issue here is the high-dimensionality of the model. You need a large number of model simulations to accurately sample the predicted behavior and have enough statistical power to confidently say that a small difference in simulation results is statistically significant.} \cite{kim2013nonlinear}. A simple example is an $n$-stable switch (for example, a cell fate model) with noisy inputs. The system will ultimately end up in one of several stable states, dependent to a large extent on a random decision process. Consider a simple bi-stable switch. Over many trials, the sampled values of the variables of this system will exhibit bi-modality, with a mean in between the two modes and a very large variance encompassing the two modes, as shown in Fig.~\ref{fig_bistable_plot}. Rather than test whether a given trajectory fits within this large variance, a more useful test would be to confirm whether the trajectory follows one of the bimodal patterns. However, complex models can contain many non-trivial features such as the aforementioned branching behavior and it is difficult to construct a generalized statistical test that encompasses all cases.

We propose a hierarchical strategy to tractably assess the replicability of WC models. This stategy takes advantage of the hierarchical nature of WC models themselves. (1) The replicability of each submodel should be statistically assesed by comparing its predicted mean and standard deviation across multiple simulators. \karrcomment{(2)}{This second point should describe statistically assessing the replicability of WC models. The main point that you're trying to get accross by proposing this hierarchical testing scheme is that its easier to debug pieces of the model first and before attemptting to debug the whole model.} \karrcomment{Functional hypothesis-based testing}{Not sure what this means} should be applied to the hybrid model as a whole. We resort to \karrcomment{testing externally measurable outcomes}{You're mixing concepts of verification and validation here which is confusing.} for a given input or perturbation. For example, the cellular doubling time is externally measurable, and should respond in a predictable way to certain gene deletions under a given set of environmental conditions. These conditions are inputs to the model and can be reliably set up for the original model as well as any re-implementations. The outputs of these high-level tests should be the same across all implementations, and can serve as a measure of reproducibility. However, a failure in these high-level tests is difficult to trace back to its origin, underscoring the importance of the subunit tests.

\section{\karrcomment{Constructing a Platform for Complex, Hybrid Modeling}{This sections needs more introduction and more clarity. I'm not sure where you're going with these thoughts. What is the purpose of the this new platform? What are its features?}}

The \textit{M. genitalium} \karrcomment{re-encoding effort}{We only briefly introduced the Rostock workshop, so I would try to avoid referring to it.} provides a useful test case for a platform designed to support large, complex, hybrid models, such as WC models. Using the outcome of the workshop as an example, we can expect the following areas to be key features in the next generation of modeling software:

\begin{itemize}
\item A hybrid timecourse solver incorporating multiple, heterogeneous codebases
\item Better standards for model integration, including how to parallelize the model for execution performance
\item Transparent encoding of model data and initialization
\end{itemize}

These focus areas require advances in simulation software, standards, and services to connect models and data. The first requirement necessitates having a way for a diverse set of solvers to communicate. There are many ways to achieve this, including shared memory buffers, memory-mapped files, or message passing. Message passing offers the additional advantage that the different solvers can reside on different machines connected by a network. Furthermore, if the structure of the data contained in the message is defined and documented, it is possible to use the message passing scheme from many different programming languages and software environments. The SBML Arrays package \cite{watanabe2016efficient} helps aid in the parallelization of certain types of models, while the Hierarchical Model Composition package \cite{smith2015sbml} offers a way to integrate models. However, there is overlap between these two packages, and their respective feature sets could be reconciled into a single scheme for encoding and simulating hierarchical models. This scheme must also address how to deal with updates to shared variables, which can be a source of conflicts and race conditions. Finally, better databases and services are needed to transparently, unambiguously tie the parameterization and initialization of the model to the data used to generate it. Linked data has proven to be extremely useful both in the annotation of XML-encoded standards \cite{novere2005minimum} and pure RDF standards \cite{galdzicki2014synthetic}.

Before a platform for hybrid modeling can be realized, the modeling community will first need to agree on a few preliminary features such a platform should have. For a hybrid solver to be useful, it should be implemented as a highly reusable software system. In so far as possible, it should be language-agnostic in order to reach the broadest possible audience. As an example, consider Geppetto, the multi-solver interface for the OpenWorm project \cite{szigeti2014openworm}. Geppetto uses the OSGi framework to interface with various solvers in a modular way. However, this system has a serious drawback: It is not language agnostic, and requires an OSGi interface written in Java for each component. This in turn imposes a burden on developers of non-Java solvers. While we are not aware of any single system that can address all of OSGi's use cases in a language agnostic way, we do acknowledge that there are libraries (such as ZeroMQ) that enable cross-process and cross-platform communication, and we believe that using communication as the starting point and building a service architecture on top of it is more language agnostic than starting with a specific service scheme (OSGi) and trying to build a system for cross-language communication.

It has been our experience that non-replicable simulations are notoriously difficult to debug. Therefore, we propose that in order to build a scalable platform, replicable simulations should be incorporated as a core feature. Since replicable simulations are difficult to achieve for hybrid models with the aforementioned synchronization issues for shared variables, we devote an entire section to this problem and how it is handled in computer science. We then suggest how these computer science techniques may be applied to timecourse simulation software.

\begin{figure*}[!t]
  \centering
  \begin{subfigure}[t]{0.5\textwidth}
    \centering
    \input{bistable-stochastic.tex}
    \caption{A timecourse plot of a stochastic bistable switch shows two distinct trajectories. The simulation was performed using Gillespie's Direct Method \cite{gillespie1977exact} implemented in libRoadRunner \cite{Somogyi17062015}}
  \end{subfigure}%
  ~
  \begin{subfigure}[t]{0.5\textwidth}
    \centering
    \begin{BVerbatim}
      -> x; 0.5 + vmax*(x^n/z^n) \
            / (15 + (x^n/z^n))
  x   ->  ; k*x/z

  z = 1*16
  x = 1.4925*16
  n = 4
  vmax = 10
  k = 2




    \end{BVerbatim}
    \caption{Antimony code for the model in part (a)}
  \end{subfigure}
  \caption{Some stochastic systems can exhibit statistical divergence, making them unsuitable for mean-and-variance tests. Each trace in part (a) corresponds to the same model with the same initial conditions and a different random seed.}
  \label{fig_bistable_plot}
\end{figure*}

\subsection{Technical Considerations for Repeatability and Reproducibility}

In this section, we describe two technical issues -- concurrency and memory layout -- which can 
cause non-replicable simulations, and outline two approaches from computer science which simulation software developers should utilize to overcome these issues.

Threading and asynchronous input/output callbacks execute instructions in an arbitrary,
non-deterministic order that is determined by a program's interaction with 
its operating environment. The push to exploit \karrcomment{multi-core}{more explicitly connect the ideas of threading, multi-core, etc. for readers not familiar with these terms} CPUs has
led computer scientists to develop strategies for deterministically running software 
in concurrent environments. C{\small ORE}D{\small ET} \cite{bergan2010coredet} and D{\small THREADS} \cite{liu2011dthreads} are two systems that have been developed to eliminate non-determinism in threading.
C{\small ORE}D{\small ET} is a framework and library for compiling C/C++ programs
such that the program behaves deterministically.
D{\small THREADS} is a replacement for the standard UNIX Pthreads library.
Both of these systems create deterministic output by preventing information sharing between
threads during a parallel phase, and deterministically allowing sharing
during a serial phase.
This biphasic structure has been shown to only impose a small overhead \cite{liu2011dthreads}. 

These libraries also eliminate non-determinism
due to variable memory layout. D{\small THREADS} achieves this through a specially implemented memory allocator. 
C{\small ORE}D{\small ET} achieves this by compiling the
memory allocator itself with the C{\small ORE}D{\small ET} framework to generate deterministic memory layout.

\karrcomment{It is important to emphasize that these tools only address determinism, not sensitivity to initial conditions.}{This paragraph may be more cofusing than helpful.}
If a model is highly sensitive to initial conditions, a small perturbation can yield significant differences in the output. This is a property of the model, not the simulation tool.
We hope that tools like C{\small ORE}D{\small ET} and D{\small THREADS} will be utilized in future simulators
to achieve replicable simulation results, and that the systems biology community should
work toward producing software with robust determinism.

\subsection{\karrcomment{Limitations of Current Standards for WC Modeling}{This section feels out of place here. This needs to be reordered so the paper flows more logically. Maybe this belongs in the introduction.}}

The \textit{M. genitalium} WC model \cite{Karr2012} consists of 28 submodels.
The submodels are implemented using a variety of modeling formalisms including
ordinary differential equations, Boolean networks, and flux balance analysis.

Although the \textit{M. genitalium} model is replicable, the model is not easily reproducible for several reasons. First,
although the model could be encoded in SBML using the Flux Balance Constraints \cite{olivier2015fbc}, Hierarchical Model Composition \cite{smith2015sbml}, Multistate and Multicomponent
Species, and Arrays \cite{watanabe2016efficient} SBML packages, there is no SBML-compatible simulator which supports all of these packages or
which can simulate multi-algorithm models. The existing SBML simulators only support 
model composition by flattening submodels into a single larger model. 

Second, the model is difficult to reproduce because the details of how experimental data was used to parameterize the \textit{M. genitalium}
are not transparent. The \textit{M. genitalium} model was parameterized using data from over 900 
sources. The WholeCellKB-MG database carefully describes these data sources \cite{karr2013wholecellkb}. Although this database is publicly accessible, 
the details of how this database was used to parameterize the model are unclear
because they are intertwined with the simulation code. 
Improved software tools must be developed to help WC modelers use the minimum information 
requested in the annotation of biochemical models (MIRIAM) \cite{novere2005minimum}
standard to annotate model components with persistent uniform resource identifiers. 

Third, the model is difficult to reproduce because its assumptions are not clearly outlined. Rather, its assumptions are
implicit in the simulation source code. This makes it difficult
to distill the model's mathematical structure. A new standard is needed to allow modelers to annotate model assumptions, including why
specific mathematical formalisms were used and why those formalisms were appropriate. This standard should reference an ontology that 
includes common modeling assumptions.

Fourth, the model is difficult to reproduce because there is not yet a consensus multi-algorithm simulation meta-algorithm 
or consensus multi-algorithm simulation software system. An accurate, scalable multi-algorithm simulation meta-algorithm
and a high performance simulation program must be developed and accepted by the modeling community. This 
simulator must have deterministic control over its execution so that it can replicate 
simulations. In particular, it must have deterministic control over its use of pseudorandom number generators. 
In addition, multithreaded simulators should use libraries like C{\small ORE}D{\small ET} and D{\small THREADS} to 
ensure deterministic behavior on multi-core systems.

\subsection{\karrcomment{Towards Hybrid Simulators}{This section doesn't add a lot to the paper. Aspects of this could be better in the introduction.}}

The \textit{M. genitalium} was constructed from a limited set of data and its creators made \karrcomment{carefully chosen compromises which reflect this fact. For example, metabolism is modeled using flux balance when a more detailed kinetic model could have been employed. Flux balance was chosen due to the fact that it does not require kinetic parameters in order to estimate the flux for a pathway, and it is difficult to obtain accurate values for these parameters in an organism such as}{This comment could be put in the introduction section.} \textit{M. genitalium}. We anticipate WC models will continue to need to make such tradeoffs in the near future, so it follows that \textit{hybrid models} will become utilized more often in modeling studies.

A hybrid model employs diverse numerical solvers in calculating a timecourse trajectory. In general, these solvers will not be united by a common theoretical ground so it will not be possible to collapse (or ``flatten'') models into a single set of differential equations or rules which govern its dynamical behavior.
Instead, the dynamics will be derived from the internal behavior of the solvers and their interactions with each other. This allows modelers the freedom of choosing the optimum solver for a given role and availability of data.
Furthermore, many models feature multiscale phenomena. In a WC model, some reactions will proceed on a faster timescale than others. The difference can be orders of magnitude, and considerable gains in performance can be realized if the fast components of the system can be simulated separately from the slow components. Indeed, SBML has attributes for such ``fast'' reactions, but complex multiscale models may have an entire hierarchy of reaction timescales and a simple true/false attribute is insufficient in this case.

Returning to the \textit{M. genitalium} model, we observe that it is divided into submodels on the basis of physiological processes (transcription, translation, metabolism, etc.). The numerical solvers for each of these submodels are all run at each timestep and the outputs are propagated to global shared variables. Assuming \textit{a priori} knowledge of the timescale of each component, we could instead subdivide the full model on the basis of timescale by putting components with similar scales in the same submodel. Furthermore, instead of running all numerical solvers at each timestep, we could employ a hierarchy of timesteps in which only the fastest submodels are run, followed by the next fastest, and so on.

Naturally, this raises theoretical questions. The behavior of complex, large-scale hybrid systems is largely unknown. Moreover, existing schemes for handling combined discrete and continuous systems often do not scale well as the number of discrete events per unit of time increases.

\section{Conclusion}

Expanded standards and new software tools are needed to facilitate replicable and reproducible WC modeling.
These standards should enable researchers to describe every model data source and assumption. One key step toward replicating WC model simulation results is expanded support among SBML-compatible simulators for the 
Flux Balance Constraints \cite{olivier2015fbc}, Hierarchical Model Composition \cite{smith2015sbml}, Multistate and Multicomponent
Species, and Arrays \cite{watanabe2016efficient} SBML packages. 
Each simulator should also have deterministic control over its use of pseudorandom number generators.
These new standards and expanded software tools will make it easier for researchers to understand, reproduce, and validate WC models. 
Lastly, continued commitment to open standards, software tools, and models will advance the reproducibility of systems biology models.

\section{Acknowledgment}

The authors thank Herbert M. Sauro for helpful discussions
on reproducibility and repeatability.

% Can use something like this to put references on a page
% by themselves when using endfloat and the captionsoff option.
\ifCLASSOPTIONcaptionsoff
  \newpage
\fi

\bibliographystyle{IEEEtran}
\bibliography{IEEEabrv,model_reproducibility}

%%%%%%%%%%%%%%%%%%%%%
%% biography section
%%%%%%%%%%%%%%%%%%%%%
\begin{IEEEbiography}[{\includegraphics[width=1in,height=1.25in,clip,keepaspectratio]{photos/medley.jpg}}]{J. Kyle Medley}
obtained his BSc in Mechanical Engineering from the University of Missouri-Kansas City and
is currently pursuing a Ph.D. in Bioengineering from the University of Washington, USA.
His research interests include modeling dynamical biophysical processes such as
metabolism and gene expression regulation.
He currently develops libRoadRunner, a software package for simulating dynamical models encoded in SBML.
\end{IEEEbiography}

\begin{IEEEbiography}[{\includegraphics[width=1in,height=1.25in,clip,keepaspectratio]{photos/karr.jpg}}]{Jonathan R. Karr}
received his Ph.D. in Biophysics and M.S. in Medicine from Stanford University, USA in 2014 and his B.S.s in Physics and Brain \& Cognitive Sciences from the Massachusetts Institute of Technology, USA in 2006. Currently, Dr. Karr is a Fellow at the Icahn School of Medicine at Mount Sinai, USA. His research focuses on the development of comprehensive WC computational models and their application to bioengineering and medicine.
\end{IEEEbiography}

\vfill

\end{document}