
%% bare_jrnl_transmag.tex
%% V1.4a
%% 2014/09/17
%% by Michael Shell
%% see http://www.michaelshell.org/
%% for current contact information.
%%
%% This is a skeleton file demonstrating the use of IEEEtran.cls
%% (requires IEEEtran.cls version 1.8a or later) with an IEEE 
%% Transactions on Magnetics journal paper.
%%
%% Support sites:
%% http://www.michaelshell.org/tex/ieeetran/
%% http://www.ctan.org/tex-archive/macros/latex/contrib/IEEEtran/
%% and
%% http://www.ieee.org/

%%*************************************************************************
%% Legal Notice:
%% This code is offered as-is without any warranty either expressed or
%% implied; without even the implied warranty of MERCHANTABILITY or
%% FITNESS FOR A PARTICULAR PURPOSE! 
%% User assumes all risk.
%% In no event shall IEEE or any contributor to this code be liable for
%% any damages or losses, including, but not limited to, incidental,
%% consequential, or any other damages, resulting from the use or misuse
%% of any information contained here.
%%
%% All comments are the opinions of their respective authors and are not
%% necessarily endorsed by the IEEE.
%%
%% This work is distributed under the LaTeX Project Public License (LPPL)
%% ( http://www.latex-project.org/ ) version 1.3, and may be freely used,
%% distributed and modified. A copy of the LPPL, version 1.3, is included
%% in the base LaTeX documentation of all distributions of LaTeX released
%% 2003/12/01 or later.
%% Retain all contribution notices and credits.
%% ** Modified files should be clearly indicated as such, including  **
%% ** renaming them and changing author support contact information. **
%%
%% File list of work: IEEEtran.cls, IEEEtran_HOWTO.pdf, bare_adv.tex,
%%                    bare_conf.tex, bare_jrnl.tex, bare_conf_compsoc.tex,
%%                    bare_jrnl_compsoc.tex, bare_jrnl_transmag.tex
%%*************************************************************************


% *** Authors should verify (and, if needed, correct) their LaTeX system  ***
% *** with the testflow diagnostic prior to trusting their LaTeX platform ***
% *** with production work. IEEE's font choices and paper sizes can       ***
% *** trigger bugs that do not appear when using other class files.       ***                          ***
% The testflow support page is at:
% http://www.michaelshell.org/tex/testflow/



\documentclass[journal,transmag,twoside]{IEEEtran}
%
% If IEEEtran.cls has not been installed into the LaTeX system files,
% manually specify the path to it like:
% \documentclass[journal]{../sty/IEEEtran}





% Some very useful LaTeX packages include:
% (uncomment the ones you want to load)


% *** MISC UTILITY PACKAGES ***
%
%\usepackage{ifpdf}
% Heiko Oberdiek's ifpdf.sty is very useful if you need conditional
% compilation based on whether the output is pdf or dvi.
% usage:
% \ifpdf
%   % pdf code
% \else
%   % dvi code
% \fi
% The latest version of ifpdf.sty can be obtained from:
% http://www.ctan.org/tex-archive/macros/latex/contrib/oberdiek/
% Also, note that IEEEtran.cls V1.7 and later provides a builtin
% \ifCLASSINFOpdf conditional that works the same way.
% When switching from latex to pdflatex and vice-versa, the compiler may
% have to be run twice to clear warning/error messages.






% *** CITATION PACKAGES ***
%
%\usepackage{cite}
% cite.sty was written by Donald Arseneau
% V1.6 and later of IEEEtran pre-defines the format of the cite.sty package
% \cite{} output to follow that of IEEE. Loading the cite package will
% result in citation numbers being automatically sorted and properly
% "compressed/ranged". e.g., [1], [9], [2], [7], [5], [6] without using
% cite.sty will become [1], [2], [5]--[7], [9] using cite.sty. cite.sty's
% \cite will automatically add leading space, if needed. Use cite.sty's
% noadjust option (cite.sty V3.8 and later) if you want to turn this off
% such as if a citation ever needs to be enclosed in parenthesis.
% cite.sty is already installed on most LaTeX systems. Be sure and use
% version 5.0 (2009-03-20) and later if using hyperref.sty.
% The latest version can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/cite/
% The documentation is contained in the cite.sty file itself.






% *** GRAPHICS RELATED PACKAGES ***
%
\ifCLASSINFOpdf
  \usepackage[pdftex]{graphicx}
  % declare the path(s) where your graphic files are
  % \graphicspath{{../pdf/}{../jpeg/}}
  % and their extensions so you won't have to specify these with
  % every instance of \includegraphics
  % \DeclareGraphicsExtensions{.pdf,.jpeg,.png}
\else
  % or other class option (dvipsone, dvipdf, if not using dvips). graphicx
  % will default to the driver specified in the system graphics.cfg if no
  % driver is specified.
  % \usepackage[dvips]{graphicx}
  % declare the path(s) where your graphic files are
  % \graphicspath{{../eps/}}
  % and their extensions so you won't have to specify these with
  % every instance of \includegraphics
  % \DeclareGraphicsExtensions{.eps}
\fi
% graphicx was written by David Carlisle and Sebastian Rahtz. It is
% required if you want graphics, photos, etc. graphicx.sty is already
% installed on most LaTeX systems. The latest version and documentation
% can be obtained at: 
% http://www.ctan.org/tex-archive/macros/latex/required/graphics/
% Another good source of documentation is "Using Imported Graphics in
% LaTeX2e" by Keith Reckdahl which can be found at:
% http://www.ctan.org/tex-archive/info/epslatex/
%
% latex, and pdflatex in dvi mode, support graphics in encapsulated
% postscript (.eps) format. pdflatex in pdf mode supports graphics
% in .pdf, .jpeg, .png and .mps (metapost) formats. Users should ensure
% that all non-photo figures use a vector format (.eps, .pdf, .mps) and
% not a bitmapped formats (.jpeg, .png). IEEE frowns on bitmapped formats
% which can result in "jaggedy"/blurry rendering of lines and letters as
% well as large increases in file sizes.
%
% You can find documentation about the pdfTeX application at:
% http://www.tug.org/applications/pdftex




% *** MATH PACKAGES ***
%
%\usepackage[cmex10]{amsmath}
% A popular package from the American Mathematical Society that provides
% many useful and powerful commands for dealing with mathematics. If using
% it, be sure to load this package with the cmex10 option to ensure that
% only type 1 fonts will utilized at all point sizes. Without this option,
% it is possible that some math symbols, particularly those within
% footnotes, will be rendered in bitmap form which will result in a
% document that can not be IEEE Xplore compliant!
%
% Also, note that the amsmath package sets \interdisplaylinepenalty to 10000
% thus preventing page breaks from occurring within multiline equations. Use:
%\interdisplaylinepenalty=2500
% after loading amsmath to restore such page breaks as IEEEtran.cls normally
% does. amsmath.sty is already installed on most LaTeX systems. The latest
% version and documentation can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/required/amslatex/math/





% *** SPECIALIZED LIST PACKAGES ***
%
%\usepackage{algorithmic}
% algorithmic.sty was written by Peter Williams and Rogerio Brito.
% This package provides an algorithmic environment fo describing algorithms.
% You can use the algorithmic environment in-text or within a figure
% environment to provide for a floating algorithm. Do NOT use the algorithm
% floating environment provided by algorithm.sty (by the same authors) or
% algorithm2e.sty (by Christophe Fiorio) as IEEE does not use dedicated
% algorithm float types and packages that provide these will not provide
% correct IEEE style captions. The latest version and documentation of
% algorithmic.sty can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/algorithms/
% There is also a support site at:
% http://algorithms.berlios.de/index.html
% Also of interest may be the (relatively newer and more customizable)
% algorithmicx.sty package by Szasz Janos:
% http://www.ctan.org/tex-archive/macros/latex/contrib/algorithmicx/




% *** ALIGNMENT PACKAGES ***
%
%\usepackage{array}
% Frank Mittelbach's and David Carlisle's array.sty patches and improves
% the standard LaTeX2e array and tabular environments to provide better
% appearance and additional user controls. As the default LaTeX2e table
% generation code is lacking to the point of almost being broken with
% respect to the quality of the end results, all users are strongly
% advised to use an enhanced (at the very least that provided by array.sty)
% set of table tools. array.sty is already installed on most systems. The
% latest version and documentation can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/required/tools/


% IEEEtran contains the IEEEeqnarray family of commands that can be used to
% generate multiline equations as well as matrices, tables, etc., of high
% quality.




% *** SUBFIGURE PACKAGES ***
%\ifCLASSOPTIONcompsoc
%  \usepackage[caption=false,font=normalsize,labelfont=sf,textfont=sf]{subfig}
%\else
%  \usepackage[caption=false,font=footnotesize]{subfig}
%\fi
% subfig.sty, written by Steven Douglas Cochran, is the modern replacement
% for subfigure.sty, the latter of which is no longer maintained and is
% incompatible with some LaTeX packages including fixltx2e. However,
% subfig.sty requires and automatically loads Axel Sommerfeldt's caption.sty
% which will override IEEEtran.cls' handling of captions and this will result
% in non-IEEE style figure/table captions. To prevent this problem, be sure
% and invoke subfig.sty's "caption=false" package option (available since
% subfig.sty version 1.3, 2005/06/28) as this is will preserve IEEEtran.cls
% handling of captions.
% Note that the Computer Society format requires a larger sans serif font
% than the serif footnote size font used in traditional IEEE formatting
% and thus the need to invoke different subfig.sty package options depending
% on whether compsoc mode has been enabled.
%
% The latest version and documentation of subfig.sty can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/subfig/



% *** FLOAT PACKAGES ***
%
%\usepackage{fixltx2e}
% fixltx2e, the successor to the earlier fix2col.sty, was written by
% Frank Mittelbach and David Carlisle. This package corrects a few problems
% in the LaTeX2e kernel, the most notable of which is that in current
% LaTeX2e releases, the ordering of single and double column floats is not
% guaranteed to be preserved. Thus, an unpatched LaTeX2e can allow a
% single column figure to be placed prior to an earlier double column
% figure. The latest version and documentation can be found at:
% http://www.ctan.org/tex-archive/macros/latex/base/


%\usepackage{stfloats}
% stfloats.sty was written by Sigitas Tolusis. This package gives LaTeX2e
% the ability to do double column floats at the bottom of the page as well
% as the top. (e.g., "\begin{figure*}[!b]" is not normally possible in
% LaTeX2e). It also provides a command:
%\fnbelowfloat
% to enable the placement of footnotes below bottom floats (the standard
% LaTeX2e kernel puts them above bottom floats). This is an invasive package
% which rewrites many portions of the LaTeX2e float routines. It may not work
% with other packages that modify the LaTeX2e float routines. The latest
% version and documentation can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/sttools/
% Do not use the stfloats baselinefloat ability as IEEE does not allow
% \baselineskip to stretch. Authors submitting work to the IEEE should note
% that IEEE rarely uses double column equations and that authors should try
% to avoid such use. Do not be tempted to use the cuted.sty or midfloat.sty
% packages (also by Sigitas Tolusis) as IEEE does not format its papers in
% such ways.
% Do not attempt to use stfloats with fixltx2e as they are incompatible.
% Instead, use Morten Hogholm'a dblfloatfix which combines the features
% of both fixltx2e and stfloats:
%
% \usepackage{dblfloatfix}
% The latest version can be found at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/dblfloatfix/




%\ifCLASSOPTIONcaptionsoff
%  \usepackage[nomarkers]{endfloat}
% \let\MYoriglatexcaption\caption
% \renewcommand{\caption}[2][\relax]{\MYoriglatexcaption[#2]{#2}}
%\fi
% endfloat.sty was written by James Darrell McCauley, Jeff Goldberg and 
% Axel Sommerfeldt. This package may be useful when used in conjunction with 
% IEEEtran.cls'  captionsoff option. Some IEEE journals/societies require that
% submissions have lists of figures/tables at the end of the paper and that
% figures/tables without any captions are placed on a page by themselves at
% the end of the document. If needed, the draftcls IEEEtran class option or
% \CLASSINPUTbaselinestretch interface can be used to increase the line
% spacing as well. Be sure and use the nomarkers option of endfloat to
% prevent endfloat from "marking" where the figures would have been placed
% in the text. The two hack lines of code above are a slight modification of
% that suggested by in the endfloat docs (section 8.4.1) to ensure that
% the full captions always appear in the list of figures/tables - even if
% the user used the short optional argument of \caption[]{}.
% IEEE papers do not typically make use of \caption[]'s optional argument,
% so this should not be an issue. A similar trick can be used to disable
% captions of packages such as subfig.sty that lack options to turn off
% the subcaptions:
% For subfig.sty:
% \let\MYorigsubfloat\subfloat
% \renewcommand{\subfloat}[2][\relax]{\MYorigsubfloat[]{#2}}
% However, the above trick will not work if both optional arguments of
% the \subfloat command are used. Furthermore, there needs to be a
% description of each subfigure *somewhere* and endfloat does not add
% subfigure captions to its list of figures. Thus, the best approach is to
% avoid the use of subfigure captions (many IEEE journals avoid them anyway)
% and instead reference/explain all the subfigures within the main caption.
% The latest version of endfloat.sty and its documentation can obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/endfloat/
%
% The IEEEtran \ifCLASSOPTIONcaptionsoff conditional can also be used
% later in the document, say, to conditionally put the References on a 
% page by themselves.




% *** PDF, URL AND HYPERLINK PACKAGES ***
%
\usepackage{url}
% url.sty was written by Donald Arseneau. It provides better support for
% handling and breaking URLs. url.sty is already installed on most LaTeX
% systems. The latest version and documentation can be obtained at:
% http://www.ctan.org/tex-archive/macros/latex/contrib/url/
% Basically, \url{my_url_here}.

\usepackage{hyperref}
\newcommand{\email}[1]{\href{mailto:#1}{#1}}

\usepackage{pdfcomment}
\newcommand{\karrcomment}[2]{\pdfmarkupcomment[markup=Highlight,color=yellow,author={Jonathan Karr}]{#1}{#2}}
\usepackage[pdftex,rgb,dvipsnames,svgnames,hyperref,table]{xcolor}
\newcommand{\change}[1]{\textcolor{DarkBlue}{#1}}
\newcommand{\changesection}{\color{DarkBlue}}
\usepackage{pgfplots}
\usepackage{caption}
\usepackage{subcaption}
\usepackage{fancyvrb}

% *** Do not adjust lengths that control margins, column widths, etc. ***
% *** Do not use packages that alter fonts (such as pslatex).         ***
% There should be no need to do such things with IEEEtran.cls V1.6 and later.
% (Unless specifically asked to do so by the journal or conference you plan
% to submit to, of course. )

\usepackage{color,soul} % for highlighting only


%http://tex.stackexchange.com/questions/61265/why-doesnt-the-ieeetran-document-class-recognize-definition-environments
\newtheorem{definition}{Definition}

% correct bad hyphenation here
\hyphenation{net-works}

% Nature News on irreproducibility:
% http://www.nature.com/news/reproducibility-1.17552

\begin{document}

\newcommand{\thetitle}{Guidelines for reproducibly building and simulating systems biology models}
\title{\thetitle}

\author{
    J. Kyle Medley$^*$ and
	Jonathan R. Karr
    
    \thanks{
        Manuscript received November 4, 2015; revised XXX XX, 2016; accepted XXX XX, 2016. Date of publication XXX XX, 2016; date of current version XXX XX, 2016.
        J. K. Medley was supported by National Institutes of Health grant R01 GM081070. J. R. Karr was supported by National Science Foundation grant 1548123 and a James S. McDonnell Foundation Postdoctoral Fellowship Award in Studying Complex Systems. The content is solely the responsibility of the authors and does not necessarily represent the views of the National Institutes of Health, the National Science Foundation, or the James S. McDonnell Foundation.
        \textit{Asterisk indicates corresponding author.}
    }
    \thanks{$^*$J. K. Medley is with the Department of Bioengineering, University of Washington, Seattle, WA 98195, USA (e-mail: \email{medleyj@uw.edu}).}
    \thanks{J. R. Karr is with the Department of Genetics \& Genomic Sciences, Icahn School of Medicine at Mount Sinai, New York, NY 10029, USA (e-mail: \email{karr@mssm.edu}).}
    \thanks{Digital Object Identifier 10.1109/TBME.XXXX.XXXXXXX}
}

% The paper headers
\markboth{IEEE Transactions on Biomedical Engineering,~Vol.~XX, No.~XX, XXXX~2016}%
{Medley and Karr: \thetitle}
% The only time the second header will appear is for the odd numbered pages
% after the title page when using the twoside option.
% 
% *** Note that you probably will NOT want to include the author's ***
% *** name in the headers of peer review papers.                   ***
% You can use \ifCLASSOPTIONpeerreview for conditional compilation here if
% you desire.




% If you want to put a publisher's ID mark on the page you can do it like
% this:
%\IEEEpubid{0000--0000/00\$00.00~\copyright~2014 IEEE}
% Remember, if you use this you must call \IEEEpubidadjcol in the second
% column for its text to clear the IEEEpubid mark.



% use for special paper notices
%\IEEEspecialpapernotice{(Invited Paper)}

\maketitle

\begin{abstract}
\textit{Objective:}
Reproducibility is the cornerstone of the scientific method. 
However, currently, many systems biology models cannot easily be reproduced.
\textit{Methods:} 
We analyzed the recent \textit{Mycoplasma genitalium} whole-cell (WC) model to determine the requirements for reproducible systems biology modeling.
\textit{Results:} 
We determined that reproducible modeling requires both repeatable model building and repeatable numerical simulation.
\textit{Conclusion:} 
New standards and simulation software tools are needed to enhance and verify the reproducibility of systems biology modeling. 
New standards are needed to explicitly document every data source and assumption, and new deterministic parallel simulation tools are needed to quickly simulate complex models.
\textit{Significance:}
We anticipate that these new standards and software will enable researchers to reproducibly build and simulate more complex models, including WC models.
\end{abstract}

\begin{IEEEkeywords}
Systems biology, whole-cell, computational modeling, reproducibility, repeatability, standards, software
\end{IEEEkeywords}

% To allow for easy dual compilation without having to reenter the
% abstract/keywords data, the \IEEEtitleabstractindextext text will
% not be used in maketitle, but will appear (i.e., to be "transported")
% here as \IEEEdisplaynontitleabstractindextext when the compsoc 
% or transmag modes are not selected <OR> if conference mode is selected 
% - because all conference papers position the abstract like regular
% papers do.
\IEEEdisplaynontitleabstractindextext
% \IEEEdisplaynontitleabstractindextext has no effect when using
% compsoc or transmag under a non-conference mode.







% For peer review papers, you can put extra information on the cover
% page as needed:
% \ifCLASSOPTIONpeerreview
% \begin{center} \bfseries EDICS Category: 3-BBND \end{center}
% \fi
%
% For peerreview papers, this IEEEtran command inserts a page break and
% creates the second title. It will be ignored for other modes.
\IEEEpeerreviewmaketitle



\section{Introduction}
% The very first letter is a 2 line initial drop letter followed
% by the rest of the first word in caps.
% 
% form to use if the first word consists of a single letter:
% \IEEEPARstart{A}{demo} file is ....
% 
% form to use if you need the single drop letter followed by
% normal text (unknown if ever used by IEEE):
% \IEEEPARstart{A}{}demo file is ....
% 
% Some journals put the first two words in caps:
% \IEEEPARstart{T}{his demo} file is ....
% 
% Here we have the typical use of a "T" for an initial drop letter
% and "HIS" in caps to complete the first word.
\IEEEPARstart{R}{eproducibility} is one of the central tenets of the scientific method.
We define \textit{reproducibility} as the ability to confirm a result via a completely independent test, including by different investigators, experimental methods, and experimental machinery. In the context of systems biology, a simulation result is \textit{reproducible} if the model which generated the result can be recreated from the scientific literature and the numerical simulation result can be regenerated from descriptions of the model and the simulation experiment.
This rigorous standard eliminates conclusions that are based on incorrect methods, machinery, or experiments, and ensures that scientific results are only accepted as facts once multiple scientists have thoroughly dismissed other potential explanations. 

We define \textit{repeatability} as the ability to regenerate a result given the same experimental machinery and conditions. In the context of systems biology, a simulation result is \textit{repeatable} if the numerical result can be regenerated from descriptions of the model and the simulation experiment.
Repeatability is a more lenient standard than reproducibility because it does not require regeneration of the model itself. 
Consequently, repeatability only eliminates scientific conclusions that are based on erroneous experiments, and cannot eliminate conclusions that are based on faulty models.

The following example illustrates the difference between reproducibility and repeatability. Consider a modeler Alice who wants to investigate a discrepant simulation result published by Bob. Bob's model predicts that knocking out regulator $Y$ causes cancer, whereas Alice's model indicates that cancer requires at least two knockouts. Fortunately, Alice can investigate Bob's results because Bob published his model and simulation experiments in standard formats. Alice uses Bob's model and simulation experiment files to \textit{repeat} Bob's simulation results, compares the two models, and finds that Bob's model generates different predictions because it uses a different rate law to describe the effect of regulator $Y$. However, because Bob's model file does not describe all of the experimental data and assumptions underlying his model, Alice cannot \textit{reproduce} Bob's rate laws and thus cannot fully resolve why their models generate different predictions.

The systems biology community, spearheaded by the Computational Modeling in Biology Network \cite{hucka2015promoting}, has developed several standard formats to exchange systems biology models and repeat systems biology simulations \cite{drager2014improving}. These formats include CellML \cite{cuellar2003overview}, the COMBINE archive \cite{COMBINE2012}, the Systems Biology Markup Language (SBML) \cite{hucka2003}, the Simulation Experiment Description Markup Language (SED-ML) \cite{sedml2011}, and the Systems Biology Graphical Notation (SBGN) \cite{LeNovereHMMSS09}. These standards support several types of systems biology models including ordinary differential equation (ODE), flux balance analysis \cite{orth2010flux}, and logical models. Numerous modeling software programs support these standards \cite{hucka2011profile}. Furthermore, several model repositories, including BioModels \cite{chelliah2015biomodels} and the CellML Model Repository \cite{lloyd2008cellml}, share models that are encoded in these standards. 

\begin{figure*}[!tb]
\centering
\includegraphics[width=\textwidth]{figure1/figure1}
% where an .eps filename suffix will be assumed under latex,
% and a .pdf suffix will be assumed for pdflatex; or what has been declared
% via \DeclareGraphicsExtensions.
\caption{Three requirements for reproducible systems biology modeling. (1) Every data source and assumption should be recorded so that each species, reaction, and rate law can be regenerated. (2) Every simulation parameter value should be recorded so that numerical simulation results can be regenerated. (3) Multiple simulation software tools should generate consistent numerical results. Reproducibility is a stricter standard than repeatability. Models are considered repeatable, but not reproducible, if they satisfy the second and third criteria.}
\label{fig_repro_diagram}
\end{figure*}

These standards and modeling software tools help researchers repeat most systems biology simulation experiments and reuse, modify, expand, and combine most systems biology models. However, these standards and software provide limited support for regenerating models themselves because they do not record all of the model design choices used to build models, including every experimental data source and assumption.

Furthermore, researchers have begun to develop more complex models which cannot be represented by the existing systems biology standards or simulated by the existing standards-compliant software. For example, several aspects of the recent whole-cell (WC) model of the gram-positive bacterium \textit{Mycoplasma genitalium} \cite{Karr2012} cannot currently be represented by SBML \cite{Waltemath2016}. In particular, SBML cannot represent the multi-algorithmic nature of the \textit{M. genitalium} WC model. In addition, no SBML-compatible simulation software program supports all of the SBML packages needed to represent WC models, including the Arrays \cite{watanabe2016efficient}, Distributions \cite{Moodie2015}, Flux Balance Constraints \cite{olivier2015fbc}, Hierarchical Model Composition \cite{smith2015sbml}, and Multistate and Multicomponent Species \cite{SBMLMulti} packages.

Here, we present a comprehensive vision for reproducible systems biology modeling, including WC modeling. First, we propose several requirements for reproducible modeling. Second, we outline several gaps in the existing standards and software for reproducible modeling and describe several new standards and software tools that are needed for reproducible modeling. Third, we describe several methods for verifying the repeatability of systems biology models.

\section{The Requirements for Reproducible Modeling}

There are three requirements for fully reproducible systems biology modeling (Fig.~\ref{fig_repro_diagram}). (1) Researchers should be able to regenerate models, including every species, reaction, and rate law, from the scientific literature. Consequently, researchers should record every data source and assumption used to build models. (2) Researchers should be able to regenerate statistically identical simulation results. Consequently, researchers should record every parameter value, algorithm, and simulation software option used to simulate models. (3) Researchers should ensure that multiple simulation software tools generate statistically identical simulation results. This is particularly helpful for identifying errors in simulation software programs. This requires standard model description formats that support all possible models, including WC models. This enables researchers to use different software programs to generate the same results.

\section{The Path to Reproducible Modeling}

Several new tools and practices are needed to help researchers reproducibly build and simulate systems biology models. (1) More comprehensive experimental databases should be developed to provide the data needed to build systems models.  For example, Karr et al. developed the WholeCellKB database \cite{karr2013wholecellkb} to organize the over 1,400 quantitative measurements used to build the \textit{M. genitalium} WC model. These databases should be machine-readable so that they can be automatically queried and used to build models. These databases should also use data models similar to that of predictive models to clarify the connection between models and their underlying experimental data. (2) New model design tools and Systems Biology Ontology (SBO) \cite{juty2013systems} terms are needed to record how models are built, including every design choice, assumption, and data source. Currently, researchers can use SBML annotations and SBO terms to record several common assumptions, such as the rapid equilibrium between free enzymes and enzyme-substrate complexes in Michaelis-Menten kinetics. However, many researchers do not utilize these annotations and the SBO does not represent all possible assumptions. Thus, researchers should develop model design tools which automatically record assumptions and data sources. The SBO should also be expanded to represent more assumptions. (3) The existing model description standards should be expanded to succinctly represent all possible models, including WC models. For example, to succinctly describe WC models, SBML should be expanded to support genomic sequences and sequence-based reaction patterns. (4) Where possible, researchers should use standard formats rather than to proprietary codes to describe models and simulation experiments. For example, researchers should use SED-ML to represent every software configuration option used to generate simulation results. (5) The existing standards-compliant simulation software programs should be expanded to support a wider variety of models. In particular, simulation software programs should be expanded to support the Arrays, Distributions, Flux Balance Constraints, Hierarchical Model Composition, and Multistate and Multicomponent Species SBML packages. Several simulation software programs, including BioUML \cite{Kolpakov2006} and iBioSim \cite{Stevens2013}, have already begun to support these packages. Unfortunately, most simulation software developers have insufficient funding to implement every package. (6) Researchers should systematically verify the statistical repeatability of every simulation software program and computational model. (7) New model verification tools should be developed to automatically identify errors in models. We developed a suite of simple tests to systematically error-check our \textit{M. genitalium} WC model. These tests checked for simple problems such as undefined species, reaction mass and charge imbalance, and inconsistent reaction rates among submodels. Nevertheless, we found this test suite invaluable for debugging our model. These tests should be generalized and new software tools should be developed to help researchers systematically evaluate such tests. (8) Every model, simulation experiment, simulation result, and simulation software program should be published open-source so that researchers have all of the information needed to reproduce reported results.

\subsection{Special Considerations for Stochastic Simulation}
Special care should be taken to ensure that stochastic simulations are repeatable because each simulation generates different results. We recommend that stochastic simulation software developers use random number generators (RNG) and seed these RNGs so that modelers can repeat results. This enables researchers to repeat not only statistical distributions, but exact trajectories, which is invaluable for identifying and debugging errors in complex models. 

Furthermore, we recommend that simulation software developers carefully review the documentation for RNG libraries and use methods which completely seed the internal state of RNGs. This recommendation comes from the observation that the \texttt{seed} method of the Python Mersenne-Twister implementation, unlike that of the C++ implementation, does not completely set the RNG's internal state. Thus, Python developers should use the RNG's \texttt{setstate} method to completely set its internal state, whereas C++ developers should use the \texttt{seed} method.

\subsection{Special Considerations for Multi-algorithm WC Models}
Multi-algorithm WC models strive to represent every gene and cell function by combining multiple submodels of individual cellular pathways, each represented using different mathematics and each trained using different experimental data \cite{Karr2015, macklin2014future, carrera2015build}. This multi-algorithm methodology is motivated by the desire to model biological systems as completely as possible, given our current knowledge, by simultaneously using fine-grained representations such as ODEs and coarse-grained representations such as Boolean networks to represent well- and poorly-characterized pathways. 

Multi-algorithm modeling is a new methodology which still lacks a strong theoretical foundation. Consequently, significant work is still needed to develop tools for building, simulating, and reproducing WC models. (1) The Hierarchical Model Composition SBML package should be extended to use the KiSAO ontology \cite{courtot2011controlled} to represent each submodel's simulation algorithm. (2) Researchers should determine how to concurrently integrate submodels that share state. Researchers are currently exploring several potential methods to concurrently integrate submodels, including shared memory and parallel discrete event simulation \cite{Goldberg2016}. (3) Researchers should develop a high-performance, reusable multi-algorithm simulator. (4) Additional tools should be developed to help researchers build and analyze WC models. (5) These programs should be integrated into a comprehensive WC modeling platform, but should be implemented as separate tools so that modelers can easily use alternative components and so that software developers can contribute to individual tools. We anticipate that this platform will enable more researchers to engage in WC modeling and accelerate the WC modeling field.

\subsection{Special Considerations for Parallel Simulation Software Programs}

Multi-threaded and multi-core programs execute individual instructions in a non-deterministic order that is determined by the program's operating environment. Multi-threaded and multi-core programs also non-deterministically allocate memory in ways that are determined by their operating environment. Special care should be taken to ensure that parallel simulation software tools are repeatable despite this non-deterministic thread execution and memory layout.

Deterministic synchronization is germane to WC models because hybrid models can exhibit the same repeatability problems as parallel computer programs. These problems occur when submodules attempt to exchange data or update shared variables. In the \textit{M. genitalium} WC model, the chromosome represents a single entity that can be accessed and modified by multiple submodules --- for example, to bind a protein to the chromosome. If two submodules attempt to bind a protein to the same locus on the chromosome, a race condition ensues. Such binding must be resolved deterministically if the simulation is to be repeatable. % Another example is utilization of a finite energy pool. At each timestep, the \textit{M. genitalium} model allocates available ATP molecules among its submodules according to each submodule's energy requirements. When the total energy requirements exceed the available reserve, a shortfall occurs and the energy resources must be partitioned among the submodules in such a way as to optimize the chances of cell survival. There are multiple strategies for performing this partitioning, but, if ranking is used to distribute energy, then care must be taken to ensure the ranking is repeatable.
We therefore briefly examine methods from computer science used to solve repeatability problems stemming from non-determinism caused by synchronization and memory layout.

\begin{figure*}[tb!]
  \centering
  \begin{subfigure}[t]{0.5\textwidth}
    \centering
    \begin{BVerbatim}
      -> x; 0.5 + vmax*(x^n/z^n) \
            / (15 + (x^n/z^n))
  x   ->  ; k*x/z

  z = 1*16
  x = 1.4925*16
  n = 4
  vmax = 10
  k = 2




    \end{BVerbatim}    
  \end{subfigure}%
  ~
  \begin{subfigure}[t]{0.5\textwidth}
    \centering
    \input{bistable-stochastic.tex}    
  \end{subfigure}
  \caption{Mean and variance testing is not an effective way to verify the repeatability of stochastic models because most simulation results cannot be uniquely specified by these statistical metrics. For example, the bistable stochastic model described by the \karrcomment{Antimony}{It would be better to describe the model rather than the Antimony code which readers won't be familiar with} code in (a) and simulated trajectories plotted in (b) cannot be completely specified by its mean and variance. Statistical equivalence should be tested using multivariate Kolmogorov-Smirnov tests. \karrcomment{Supplementary Material S1 contains all of the Python code needed to reproduce this example.}{TODO}}
  \label{fig_bistable_plot}
\end{figure*}

Several libraries have been developed to eliminate non-determinism from multi-threaded programs. These include C{\small ORE}D{\small ET} \cite{bergan2010coredet} and D{\small THREADS} \cite{liu2011dthreads}.
C{\small ORE}D{\small ET} is a library for compiling C/C++ programs such that they behave deterministically.
D{\small THREADS} is a replacement for the standard UNIX Pthreads library.
Both systems create deterministic programs by preventing information sharing between
threads during parallel phases, and deterministically allowing sharing during serial phases.
The systems also eliminate non-determinism due to variable memory layout. D{\small THREADS} achieves this by implementing a special deterministic memory allocator. 
C{\small ORE}D{\small ET} deterministically allocates memory compiling memory allocators with the C{\small ORE}D{\small ET} library to generate deterministic memory allocators.
C{\small ORE}D{\small ET} and D{\small THREADS} have both been shown to impose minimal overhead \cite{liu2011dthreads}. We recommend that simulation software developers use C{\small ORE}D{\small ET} or D{\small THREADS} to create repeatable multi-threaded simulation software programs.

Computer scientists have also developed several strategies to deterministically execute distributed programs, including log-based approaches which enable repetition by recording the order of every operation, message ordering approaches which require individual operations to specify unique execution times, repeatable mechanisms for breaking ties between simultaneous operations, priority fields, virtualization, and hybrid approaches \cite{mclean2000repeatability, fujimoto1997zero, patil2010pinplay}. We recommend that simulation software developers use these approaches to create repeatable distributed simulation software programs.

\subsection{Special Considerations for Chaotic Systems}

Special care should also be taken to repeat the simulation results of chaotic models which are highly sensitive to their initial conditions. \karrcomment{We recommend that modelers communicate initial conditions and parameters with sufficient numerical precision that chaotic model simulation results can be repeated.}{What else can we say?}

\section{Methods for Verifying Repeatability}
After adopting the methods outlined above to repeatably simulate models, we recommend that systems biology modelers and software developers verify repeatability by testing the statistical equivalence of simulation results generated by multiple simulation software programs for the same model, and then correcting any discrepancies identified by this analysis. Bergmann and Sauro \cite{bergmann2008comparing} and Evans et al. \cite{evans2008sbml} have used this approach to assess the repeatability of models across several simulators, identifying errors in several simulation programs.

Model checking and unit testing are two formalisms that can be used to verify repeatability. Simulation-based model checkers, such as SPIN \cite{holzmann1997model}, are extensively used in electrical engineering to verify the behavior of models. Simulation-based checkers execute multiple simulations and then verify the statistical properties, such as the mean and variance, of their results \cite{kwiatkowska2011prism}. Model verification systems could be adopted to verify the repeatability of systems biology models. Unit testing is a powerful strategy for verifying the behavior of software. Unit testing could also be used to verify the repeatability of systems biology models and simulation programs by (1) using multiple simulators to execute multiple simulations and (2) comparing their predicted means and variances.

\subsection{Special Considerations for Stochastic Models}
In theory, the repeatability of stochastic models can be assessed by verifying the statistical equivalence of simulation results across multiple simulators. Statistical equivalence should be tested using multivariate Kolmogorov-Smirnov tests \cite{justel1997multivariate}. Tests which only check the predicted mean and variance are insufficient because most model predictions cannot be fully specified by their mean and variance. For example, the bistable system illustrated in Fig.~\ref{fig_bistable_plot} cannot be specified by its mean and variance. Furthermore, verifying repeatability with high confidence requires large numbers of simulations for models with rare events \cite{kim2013nonlinear}. 

In general, it is difficult to verify the repeatability of complex algorithms by testing the statistical equivalence of their numerical results across multiple simulators. In addition to testing the statistical equivalence of simulation results across multiple simulators, we recommend that modelers and simulation software developers (1) utilize the methods described in the previous section to construct deterministic simulators and (2) verify that each simulation software program can exactly regenerate its own numerical results. This approach facilitates the debugging of complex simulation algorithms.

\subsection{Special Considerations for Hierarchical Models}
The repeatability of hierarchical models, such as WC models, can be efficiently verified by taking advantage of their hierarchical nature. (1) Modelers should verify the repeatability of each submodel. (2) Modelers should verify the repeatability of the combined model. The main advantage of this hierarchical strategy is that it facilitates debugging by breaking models into smaller, more easily testable units. 

\subsection{Special Considerations for Multi-algorithm WC Models}
As introduced above, multi-algorithm modeling is a new methodology which still requires significant fundamental research. Thus, multi-algorithm simulation algorithms will need to be extensively tested to verify their repeatability, as well as to understand their fundamental properties such as how their runtime scales with the number of submodels and how the variance in their predicted dynamics varies with the simulation timestep.

\section{Conclusion}

Expanded standards and simulation software tools are needed to enhance the reproducibility of systems biology modeling. These standards and software tools should enable researchers to regenerate models from our scientific knowledge by recording every model design decision, assumption, data source, parameter, and software option. These software tools should also enable researchers to regenerate simulation results by using pseudo-random number generators and deterministic multi-threading libraries. These standards and software tools should also support all possible systems biology models, including WC models. This requires expanded support among simulation software programs for the Arrays, Distributions, Flux Balance Constraints, Hierarchical Model Composition, and Multistate and Multicomponent Species SBML packages. These simulation software tools also should be extensively tested to verify that they produce consistent simulation results. Lastly, we believe that continued commitment to open standards, software tools, and models is critical to advancing the reproducibility of systems biology modeling.

\section{Acknowledgment}

The authors thank Herbert Sauro for helpful discussions
on reproducibility and repeatability.

% Can use something like this to put references on a page
% by themselves when using endfloat and the captionsoff option.
\ifCLASSOPTIONcaptionsoff
  \newpage
\fi

\bibliographystyle{IEEEtran}
\bibliography{IEEEabrv,model_reproducibility}

%%%%%%%%%%%%%%%%%%%%%
%% biography section
%%%%%%%%%%%%%%%%%%%%%
\begin{IEEEbiography}[{\includegraphics[width=1in,height=1.25in,clip,keepaspectratio]{photos/medley.jpg}}]{J. Kyle Medley}
obtained his BSc in Mechanical Engineering from the University of Missouri-Kansas City and
is currently pursuing a Ph.D. in Bioengineering from the University of Washington, USA.
His research interests include modeling dynamical biophysical processes such as
metabolism and gene expression regulation.
He currently develops libRoadRunner, a software package for simulating dynamical models encoded in SBML.
\end{IEEEbiography}

\begin{IEEEbiography}[{\includegraphics[width=1in,height=1.25in,clip,keepaspectratio]{photos/karr.jpg}}]{Jonathan R. Karr}
received his Ph.D. in Biophysics and M.S. in Medicine from Stanford University, USA in 2014 and his B.S.s in Physics and Brain \& Cognitive Sciences from the Massachusetts Institute of Technology, USA in 2006. Currently, Dr. Karr is a Fellow at the Icahn School of Medicine at Mount Sinai, USA. His research focuses on the development of comprehensive WC computational models and their application to bioengineering and medicine.
\end{IEEEbiography}

\vfill

\end{document}
